arangod/Cluster/FollowerInfo.cpp

```diff
@@ -246,46 +157,187 @@ Result FollowerInfo::remove(ServerID const& sid) {
+    return agencyRes;
+  }
+  if (agencyRes.is(TRI_ERROR_CLUSTER_NOT_LEADER)) {
+    // TODO: Do we need to rollback?
+    return agencyRes;
+  }
 
+  // rollback:
+  _followers = oldFollowers;
+  _failoverCandidates = oldFailovers;
+  std::string errorMessage =
+      "unable to remove follower from agency, timeout in agency CAS operation "
+      "for key " +
+      _docColl->vocbase().name() + "/" + std::to_string(_docColl->planId()) +
+      ": " + TRI_errno_string(agencyRes.errorNumber());
+  LOG_TOPIC("a0dcc", ERR, Logger::CLUSTER) << errorMessage;
+  agencyRes.resetErrorMessage<std::string>(std::move(errorMessage));
+  return agencyRes;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief clear follower list, no changes in agency necessary
+//////////////////////////////////////////////////////////////////////////////
+
+void FollowerInfo::clear() {
+  WRITE_LOCKER(canWriteLocker, _canWriteLock);
+  WRITE_LOCKER(writeLocker, _dataLock);
+  _followers = std::make_shared<std::vector<ServerID>>();
+  _failoverCandidates = std::make_shared<std::vector<ServerID>>();
+  _canWrite = false;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief check whether the given server is a follower
+//////////////////////////////////////////////////////////////////////////////
+
+bool FollowerInfo::contains(ServerID const& sid) const {
+  READ_LOCKER(readLocker, _dataLock);
+  auto const& f = *_followers;
+  return std::find(f.begin(), f.end(), sid) != f.end();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief Inject information of a insync followers that we knew about
+///        before a failover to this server has happened
+////////////////////////////////////////////////////////////////////////////////
+void FollowerInfo::insertFollowersBeforeFailover(std::vector<std::string> const& previousInsyncFollowers) {
+  // This function copies over the information taken from the last CURRENT into a local vector.
+  // Where we remove the old leader and ourself from the list of followers
+  WRITE_LOCKER(writeLocker, _dataLock);
+  TRI_ASSERT(_failoverCandidates != nullptr && _failoverCandidates->empty());
+  if (previousInsyncFollowers.size() > 1) {
+    auto ourselves = arangodb::ServerState::instance()->getId();
+    auto failoverCandidates =
+        std::make_shared<std::vector<ServerID>>(previousInsyncFollowers);
+    auto myEntry =
+        std::find(failoverCandidates->begin(), failoverCandidates->end(), ourselves);
+    // We are a valid failover follower
+    TRI_ASSERT(myEntry != failoverCandidates->end());
+    // The first server is a different leader! (For some reason the job can be
+    // triggered twice) TRI_ASSERT(myEntry != failoverCandidates->begin());
+    failoverCandidates->erase(myEntry);
+    // Put us in front, put old leader somewhere, we do not really care
+    _failoverCandidates = failoverCandidates;
```

Comment:
Yes, this is a bit misleading.
The Follower here, are the real followers (excluding ourself)
When writing information to the agency there is only one array of servers.
The first is the LEADER (ourself in this case) all others are followers.
=> We remove us, and keep the old leader in second place, in case it comes back soonish